src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -618,31 +623,17 @@ struct overwrite_plan_t {
    * original extent into at most three parts: origin-left, part-to-be-modified
    * and origin-right.
    */
-  void evaluate_operations(Transaction& t) {
+  void evaluate_operations() {
     auto actual_write_size = get_pins_size();
     auto aligned_data_size = get_aligned_data_size();
     auto left_ext_size = get_left_extent_size();
     auto right_ext_size = get_right_extent_size();
 
-    auto can_merge = [](Transaction& t, paddr_t paddr) {
-      CachedExtentRef ext;
-      if (paddr.is_relative() || paddr.is_delayed()) {
-	  return true;
-      } else if (t.get_extent(paddr, &ext) ==
-	Transaction::get_extent_ret::PRESENT) {
-	// FIXME: there is no need to lookup the cache if the pin can 
-	// be associated with the extent state
-	if (ext->is_mutable()) {
-	  return true;
-	}
-      }
-      return false;
-    };
     if (left_paddr.is_zero()) {
       actual_write_size -= left_ext_size;
       left_ext_size = 0;
       left_operation = overwrite_operation_t::OVERWRITE_ZERO;
-    } else if (can_merge(t, left_paddr)) {
+    } else if (!is_left_stable) {
```

Comment:
It looks like we also need to consider INITIAL_WRITE_PENDING state here. Am I missing something?

Reply:
See the removed `can_merge()` logic that `paddr.is_relative()` will return `true` and cause merging.

IIUC `INITIAL_WRITE_PENDING` implies `paddr.is_relative()`.

`INITIAL_WRITE_PENDING` isn't stable and will also cause merging in the new logic.

Reply:
But the unit tests are failed, so probably the test cases are not accurate, or something else is missing here.

Reply:
got it

Reply:
I still have one question here. In RBM, the extent's address is not `relative` at `INITIAL_WRITE_PENDING` state because it directly allocates the blocks from RBM, right?

Reply:
@myoungwon Right, this is why I think checking the absoluteness of the address is backend dependent and looks confusing here.